vllm/model_executor/custom_op.py

```diff
@@ -1,52 +1,106 @@
+import torch
 import torch.nn as nn
 
+import vllm.envs as envs
 from vllm.utils import is_cpu, is_hip, is_tpu, is_xpu
 
 
 class CustomOp(nn.Module):
 
     def __init__(self, *args, **kwargs):
         super().__init__()
+        self._is_compiled = False
         self._forward_method = self.dispatch_forward()
 
     def forward(self, *args, **kwargs):
         return self._forward_method(*args, **kwargs)
 
-    def forward_native(self, *args, **kwargs):
+    @staticmethod
+    def forward_static(*args, **kwargs):
         """PyTorch-native implementation of the forward method.
 
-        This method is optional. If implemented, it can be used with compilers
-        such as torch.compile or PyTorch XLA. Also, it can be used for testing
-        purposes.
+        This method is optional. If implemented, it can be used with
+        `torch.compile` to generate the custom kernel for the op.
+        This function is necessary because any method that uses `self`
+        triggers re-compilation in `torch.compile`.
         """
         raise NotImplementedError
 
+    def forward_native(self, *args, **kwargs):
+        """PyTorch-native implementation of the forward method.
+
+        This method is expected to invoke the forward_static method
+        with the input arguments and the attributes of `self`.
+        By default, we assume that forward_static does not need any
+        attributes of `self`.
+        """
+        return self.forward_static(*args, **kwargs)
+
+    def forward_compile(self, *args, **kwargs):
+        """Runs a torch-compiled version of forward_native.
+
+        This method hooks into the forward_native method and compiles
+        `forward_static` using `torch.compile` if it has not been compiled yet.
+        """
+        if not self._is_compiled and not envs.VLLM_TEST_DYNAMO_GRAPH_CAPTURE:
```

Comment:
I think this env var does not matter here? iirc, @anijain2305 said that `torch.compile` will automatically ignore inner function that is compiled, and try to inline the original function.

Reply:
Yeah it might be ok to remove it. I just wanted to avoid any potential unnecessary issues when `torch.compile` is applied twice.

Reply:
BTW, is there any flag that we can use to check whether the model is being compiled or not (like `is_capturing` API in CUDA graph)? If so, I'd like to use that instead of this flag.

Reply:
You can use `torch._utils.is_compiling()` to check if something is being compiled.

Reply:
@bnellnm Thanks! updated the PR.